### ISSUE TITLE
merge llvm.sideeffect change

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/compiler-team#177]
# Meeting proposal info

- **Title:** merge llvm.sideeffect change
- **Type:** technical
- **Proposed by:** @nikomatsakis 
- **Link:** (inline, see below)
- **Scheduled dates for discussion:**
  - Not yet scheduled, we'll add content here when it gets scheduled

# Proposal details

Rust has had a long-standing problem where LLVM optimizes empty loops into UB. This can very easily create real bugs for folks. We've got a pending PR (#59546) that fixes this by inserting `llvm.sideeffect` instructions, but until recently it was blocked on getting some lolbench measurements done. These measurements were [recently completed] (thanks @mati865!). They suggest the impact is "generally" small but non-zero (I put them in a [world readable spreadsheet](https://docs.google.com/spreadsheets/d/1I5p0_ChO7xakIYUvbhVlgmaKnvDdvOnkpv7zDZAy-CQ/edit?usp=sharing) here, feel free to copy and play with it). We need to make a decision whether this is good enough! I'm going to cc the @rust-lang/lang team since I think lang team input would be useful here.

[recently completed]: https://github.com/rust-lang/rust/pull/59546#issuecomment-532752661

# About this issue

This issue corresponds to a meeting proposal for the compiler team
[steering meeting][]. It corresponds to a possible topic of
discussion. You can read more about the [steering meeting procedure
here][proc].

# Comment policy

These issues are meant to be used as an "announcements channel"
regarding the proposal, and not as a place to discuss the technical
details. Feel free to subscribe to updates. We'll post comments when
reviewing the proposal in meetings or making a scheduling decision.
In the meantime, if you have questions or ideas, ping the proposers
on [Zulip] (or elsewhere).

[Zulip]: https://github.com/rust-lang/compiler-team/blob/master/about/chat-platform.md
[steering meeting]: https://github.com/rust-lang/compiler-team/blob/master/about/steering-meeting.md
[proc]: https://github.com/rust-lang/compiler-team/blob/master/procedures/steering-meeting.md
[this template]: https://github.com/rust-lang/compiler-team/blob/master/proposals/_template.md
[hackmd]: https://hackmd.io/
[gist]: https://gist.github.com/

**Comment by nikomatsakis** [comment]
Discussed in today's [design meeting](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/design.20meeting.202019-09-27/near/176749849). We opted not to schedule this but instead:

- to merge the PR with a `-Z` flag
- and to get some measurements from real-world applications
    - @michaelwoerister will measure impact on FF
    - @nikomatsakis will find someone to measure impact on some async-related applications

**Comment by nikomatsakis** [comment]
Discussed again. The PR has landed, but we've not gathered the follow-up data we meant to. We decided to try and force ourselves a bit into action. The plan is to enable this by default in ~1 month, but we have a bit of time to do some measurements beforehand.

One strategy for landing would be to keep the `-Z` flag for some time but have it be used to restore the old behavior, to help make it easier to do measurements.

**Comment by michaelwoerister** [comment]
Unless I messed something up in my tests, the change does not seem to have any real affect on Firefox performance:
https://treeherder.mozilla.org/perf.html#/compare?originalProject=try&originalRevision=2b9e4386bfc5de74883781a869b95521e5a36290&newProject=try&newRevision=9e5d8cb0607a10a3645d5d4ac79f2df9864d41d7&framework=1

**Comment by Mark-Simulacrum** [comment]
We decided that it would likely be fine (especially given firefox results) to go ahead with the change, and potentially reconsider if we get significant feedback once the change rolls out. I will be working on drafting a blog post (probably for inside rust, but maybe even the main blog, not yet clear) announcing the change and describing what it does and what impact we expect it to have.

**Comment by Amanieu** [comment]
Does this mean that we can now disable the use of TrapUnreachable in LLVM? (#45920)

**Comment by rkruppe** [comment]
`TrapUnreachable` also functions as a mitigation for programs that fall off the end of a function due to legitimate UB in the program, so I don't think fixing infinite loops means we can/should just disable `TrapUnreachable`.

**Comment by joshtriplett** [comment]
It would be nice to have an option to disable TrapUnreachable, for applications that need to squeeze out every last extraneous byte.

**Comment by nikomatsakis** [comment]
This discussion should probably take place on rust-lang/rust##6402 or in some other context.

**Comment by Mark-Simulacrum** [comment]
The PR implementing this change uncovered that while the implementation gated behind -Zinsert-sideeffect does work, it has a large performance regression for [compile times](https://perf.rust-lang.org/compare.html?start=66b0c97070f422cb82baaaafc79ee94cab4396c5&end=548b5e75afd6bad696920dfdb69c9812ce0488f1). This is likely due to generating tons of side effect calls (at least one per function entry), which LLVM today also does not collapse into one after inlining. That generates horribly bloated IR, likely slowing down LLVM passes: https://github.com/rust-lang/rust/pull/68402#issuecomment-576810360.

I believe that the compile time regression is *not* because rustc itself is slower, i.e., there is still little to no runtime effect. However, I don't think we can land the patch as-is -- asking everyone to pay 5-30% worse compile times for a relatively edge case UB mitigation is probably not tenable, particularly as it feels like there should be better options out there.

I believe it's worth discussing what approach we want to take here. There's some discussion on rust-lang/rust#68402, too, that is plausibly worth reading.

I believe that in order to be 100% sure that the UB problem is solved, we essentially need to emit this number of side effect calls. If we start to try and narrow that down while still preserving all cases, I think we're going to end up in a situation where we're basically just reimplementing LLVM (i.e., we need to always keep up with its optimizations). Now, of course, the more pessimistic we are, the less work there is, but also compile times are worse.

Based on that, it's my impression that it would be worth landing some basic mitigations in the near term. In particular, I would suggest that we make sure the trivial `loop {}` case compiles correctly via side effect, but do not try to do more than that initially.

I think fully resolving the bug (https://github.com/rust-lang/rust/issues/28728) will need upstream LLVM changes as I believe have been suggested, though not implemented, which enable rustc to tell LLVM that Rust functions do *not* have forward progress guarantees.

I think the design meeting, or whatever process we choose, is going to primarily have to answer these questions:
 * How much of a compile time hit are we willing to eat in favor of soundness?
   * e.g., the function-local variant of the patch regresses only [3-7%](https://perf.rust-lang.org/compare.html?start=ae66171b41b1d05128afd0d666475bc4ba617469&end=d2668446195ecb00a2bf5584187584d40ec46e7d).
 * Do we want to dedicate developer resources to a global analysis that aims to clean up the side effects we emit at function boundaries? Are we okay with such a global analysis?

I am not sure myself that a design meeting is the best way to do this, but I also think that we're unlikely to find the time to answer these questions without such a meeting.

**Comment by nikomatsakis** [comment]
> Based on that, it's my impression that it would be worth landing some basic mitigations in the near term. In particular, I would suggest that we make sure the trivial loop {} case compiles correctly via side effect, but do not try to do more than that initially.

I'm curious whether you considered the case of a function that directly calls itself? This is the only case that's actually bitten me in practice, and it's fairly easy to do (in particular, there are times when one wants a wrapper that goes from e.g. an impl on `&T` to an `impl` on `T`, and you wind up writing something like `(**self).method()`, and if you get the number of `*` wrong, you're in trouble; this is precisely why I now write `T::method(self, ...)` to avoid this kind of error, but not everyone has that practice).

**Comment by nikomatsakis** [comment]
We discussed this in our [planning meeting today](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/planning.20meeting.202020.2E02.2E14/near/188217196). The conclusion was that scheduling a meeting on this would not yet be productive, but that a good next step would be to draft a blog post summarizing:

* What is the problem
* What efforts have been done to resolve it
* What is the impact of those efforts on compillation time

which we can then use to circulate amongst LLVM devs and try to get some ideas for better directions, and/or at least capture the status of our efforts. @Mark-Simulacrum has volunteered to try and draft that post. 🎉

**Comment by nikomatsakis** [comment]
In the planning meeting today we decided to close this meeting proposal for now. We don't have much of an agenda or topics that still need to be discussed as a team that are specific to this issue.
